### ISSUE TITLE
Support for custom conditional exports in npm specifier

### ISSUE THREAD

**Issue by nestarz** [denoland/deno#23757]
### What is the problem this feature would solve?
Deno currently does not provide a way to specify custom export conditions when importing npm packages. This makes it **impossible** to import non-default exports that are conditionally exported based on certain conditions, such as the "react-server" condition used by React Server Components.

See: 
https://nodejs.org/api/packages.html#conditional-exports
https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md

### What is the feature you are proposing to solve the problem?
Add support for a config flag in Deno, similar to the one supported by Node.js and Bun. This flag would allow specifying custom export conditions when importing npm packages.

For example, in Bun, thanks to this issue https://github.com/oven-sh/bun/issues/8990 you can do:
```
bun --conditions special-condition
```

Then, if you import an npm package, it will follow the conditions specified in the package's `exports` field.

Ideally, it would be great to set up this condition inside the `deno.json` config, so that it can be used by Deno Deploy as well. An other option would be to specify it directly in the npm specifier, so that no config file are needed ?

### What alternatives have you considered?
One workaround would be to directly import the specific file path, like:
```typescript
import "npm:/react@19.0.0-beta-b498834eab-20240506/cjs/react.react-server.development.js"
```

However, this doesn't work if the file is not explicitly exported in the package's `exports` field:
```
Uncaught TypeError: Failed resolving package subpath './cjs/react.react-server.development.js' for '/Users/Library/Caches/deno/npm/registry.npmjs.org/react/19.0.0-beta-b498834eab-20240506/package.json'
  Caused by:
    [ERR_PACKAGE_PATH_NOT_EXPORTED] Package subpath './cjs/react.react-server.development.js' is not defined by "exports" in '/Users/Library/Caches/deno/npm/registry.npmjs.org/react/19.0.0-beta-b498834eab-20240506/package.json' imported from '/Users/code/$deno$repl.ts'
```

I think this is intentional from the Deno team to not expose what's not exposed through the `exports` option. But then, Deno should provide a way to define the export conditions like Bun and Node does.

**Comment by terrablue** [comment]
I would welcome such a change. It would be particularly useful during compilation using `deno compile`, in order to tap into a `runtime` import condition which only imports packages/package paths needed during runtime. A prominent example is Svelte: excluding `svelte/compiler` during runtime is only possible with export conditions, and that's how we currently do it with Bun.

**Comment by kigawas** [comment]
I've observed the following behavior

> When there is a package A with exports deno condition, if we import A directly with deno, deno export works. If we import package B using package A, deno export will not work

https://github.com/denoland/deno/discussions/17964#discussioncomment-10917259

**Comment by davesteinberg** [comment]
I seem to have hit a dead-end trying to use Deno, Drizzle, and libsql together. The issue is that libsql uses a conditional export to determine whether to export its node or web client by default, and for "deno", you get the less capable web client, even though Deno can run the node client just fine. I've described the problem in more detail [here](https://github.com/drizzle-team/drizzle-orm/discussions/3122). I don't know if Drizzle might provide a solution, but it seems to me that being able to specify to customize the export condition would solve it from the Deno side.

**Comment by davesteinberg** [comment]
After looking at conditional exports some more, I suppose allowing _additional_ conditions to be specified wouldn't actually solve my problem, since there is no higher priority condition in libsql's export than "deno" that could be used to export the node client. 

I would need to be able to _remove_ the default "deno" condition, or some change would be required on the libsql side, as well, such as adding a higher priority "node-override" condition.

**Comment by terrablue** [comment]
@davesteinberg The semantics of custom conditional exports somewhat compete with the semantic of specific runtime keys. In custom exports, the idea is to buy into a certain feature (say, in a compile vs run situation), whilst runtime keys indicate what is supported for the given runtime. The way I read it, libsql would need to signal it fully supports Deno by switching from `web` to `node` for Deno. Given that Deno's NPM compat is only somewhat recent, it isn't surprising their default is web, but it might merit a refresh with Deno 2.

**Comment by dmint789** [comment]
What is the current status of this issue? This missing feature is currently the only reason I still have to have Node/NPM installed to be able to work with my Next JS project :(

**Comment by bartlomieju** [comment]
This will be shipped in Deno v2.3.6 later this week and available as the `--unstable-node-conditions` flag. I would appreciate trying out once it lands and letting us know if there are any troubles. We'll try to stabilize it for Deno v2.4.

**Comment by bartlomieju** [comment]
`deno upgrade canary` will get you to a version that supports this. Give it a go and let us know what you think.

**Comment by mintydev789** [comment]
@bartlomieju I'm getting `Warning 'node-conditions' isn't a valid unstable feature` despite having upgraded to the latest canary release. I just added this to my `deno.jsonc`:

```json
  "unstable": ["node-conditions"]
```

**Comment by bartlomieju** [comment]
> [@bartlomieju](https://github.com/bartlomieju) I'm getting `Warning 'node-conditions' isn't a valid unstable feature` despite having upgraded to the latest canary release. I just added this to my `deno.jsonc`:
> 
>   "unstable": ["node-conditions"]

@mintydev789 the way to use the flag is `deno run --unstable-node-conditions=<conditions...> main.js`, you can't specify it in the config file - the flag needs to receive at least one value.

**Comment by mintydev789** [comment]
> > [@bartlomieju](https://github.com/bartlomieju) I'm getting `Warning 'node-conditions' isn't a valid unstable feature` despite having upgraded to the latest canary release. I just added this to my `deno.jsonc`:
> > "unstable": ["node-conditions"]
> 
> [@mintydev789](https://github.com/mintydev789) the way to use the flag is `deno run --unstable-node-conditions=<conditions...> main.js`, you can't specify it in the config file - the flag needs to receive at least one value.

Ohh, of course. But how do I use this with `deno task`? Or is that not possible?

**Comment by mintydev789** [comment]
@bartlomieju ohh, got it now. So, now I have this command: `deno run --allow-env --allow-sys --allow-read --allow-run --unstable-node-conditions=react-server npm:drizzle-kit push`, but it still doesn't work. I'm getting the error `Error: This module cannot be imported from a Client Component module. It should only be used from a Server Component.` So the conditions flag isn't making a difference here. For reference, this is the command I'm trying to replicate: `NODE_OPTIONS=\"--conditions=react-server\" npx drizzle-kit push`.

**Comment by bartlomieju** [comment]
I'll have to debug it, maybe there's a subprocess being spawned and the flag is not getting forwarded properly.

**Comment by mintydev789** [comment]
> I'll have to debug it, maybe there's a subprocess being spawned and the flag is not getting forwarded properly.

I can provide an example, if you need it.

**Comment by bartlomieju** [comment]
@mintydev789 please do, that would be helpful

**Comment by mintydev789** [comment]
@bartlomieju Steps to reproduce:

1. Clone [this branch](https://github.com/cubing/cubingcontests.com/tree/auth) and `cd` into it (it has to be the `auth` branch)
2. `cp .env.example .env && cp .env.example client/.env.local`
3. `docker compose up -d`
4. `cd client && deno install --allow-scripts`
5. `deno task db:studio`

**Comment by kigawas** [comment]
> I've observed the following behavior
> 
> > When there is a package A with exports deno condition, if we import A directly with deno, deno export works. If we import package B using package A, deno export will not work
> 
> [#17964 (reply in thread)](https://github.com/denoland/deno/discussions/17964#discussioncomment-10917259)

@bartlomieju This still doesn't work. You can test with this js file (see comments inside):

https://github.com/ecies/js/blob/master/example/runtime/main.js#L8

**Comment by bartlomieju** [comment]
@mintydev789 your scenario now works with #29640. @kigawas I'm still debugging the scenario you provided.

**Comment by mintydev789** [comment]
> [@mintydev789](https://github.com/mintydev789) your scenario now works with [#29640](https://github.com/denoland/deno/pull/29640). [@kigawas](https://github.com/kigawas) I'm still debugging the scenario you provided.

I tested it and I think that works, thanks! I just can't use the latest canary due to [another issue](https://github.com/denoland/deno/issues/29645) now, haha.

**Comment by nestarz** [comment]
@bartlomieju is there a way to try this in Deno Deploy (EA) - which have the 2.3.6 version ?

**Comment by bartlomieju** [comment]
@nestarz not yet, keep an eye on https://github.com/denoland/deno/pull/29848

**Comment by kigawas** [comment]
I confirmed that `deno run --unstable-node-conditions deno` works with `deno` conditional exports on latest canary (2.3.7+4a0cd18), thanks to @bartlomieju

**Comment by nestarz** [comment]
@bartlomieju Deno Deploy EA can't set DENO_* variables, could you fill an exception for this flag ?
